devicemapper: Compare uuid of base device on startup

rhvgoyal opened the pull request:
Fixes issue #13894 

It is easy for one to use docker for a while, shut it down and restart
docker with different set of storage options for device mapper driver
which will effectively change the thin pool. That means any of the
metadata stored in /var/lib/docker/devicemapper/metadata/ is not valid
for the new pool and user will run into various kind of issues like
container not found in the pool etc.

Users think that their images or containers are lost but it might just
be the case of configuration issue. People might use wrong metadata
with wrong pool.

To detect such situations, save UUID of base image and once docker
starts later, query and compare the UUID of base image with the
stored one. If they don't match, fail the initialization with the
error that UUID failed to match.

That way user will be forced to cleanup /var/lib/docker/ directory
and start docker again.

Signed-off-by: Vivek Goyal vgoyal@redhat.com

rhvgoyal commented:
@vbatts What do you think about this one. I think with this we should be able to catch lot of errors early and tell users that there was most likely a configuration issue.

unclejack commented:
I haven't reviewed the actual code, but it sounds like the right thing to do.

vbatts commented:
so, this is one check. I don't think this will go the full way for a configuration/state comparison.

This kind of check may be nicer to make a part of a generic call around `Init()`. Like `LoadConfig(homePath string, configOpts map[string]string)` or similar. And have a pre-defined error in the `graphdriver` package like `ErrConfigChanged` or `ErrDeviceChanged`, which should expect to run a `Validate(path, opts)` on the setup, before continuing.

It seems like a generic case like this, for all drivers, could allow for catching odd changes with config, or switched pool devices or similar.

Thoughts?

Otherwise, this PR is simple and overall LGTM. I'll walk through the code more in a minute.

rhvgoyal commented:
Right now validation is practically in Init(). Just 2 levels down. And Init() is right place to do such kind of checks for every graph driver.

What other cases of error do you want to detect/cover? For devicemapper, I think it should cover
vast majority of issues. If there are more, we can add more checks down the line.

vbatts commented:
agreed `Init()` is a right place for it. Besides the thin pool being different, I could imaging block size changes or similar needing to bail. Or on btrfs, if the the device is changed, a Check or Validate would include asserting images/container subvolumes are present.

vbatts commented:
This can all stay in Init(), it seems like the pattern for "what is involved in Init of a storage driver?" is what i'm looking at.

rhvgoyal commented:
@vbatts 

I suspect that block size will be stored in pool metadata and if you try to create same pool with different block size this time, pool creation should fail. Otherwise all the metadata thin pool has already created will not work.

I think every target will have separate sanity checks based on kind of tunables they offer and they all need to come up with their own checks to make sure they have right configuration. But I am open to any suggestions to improve it.

vbatts commented:
right on. For now we'll review this PR as is, and let the rest evolve.

vbatts commented on daemon/graphdriver/devmapper/deviceset.go:
```diff
@@ -669,9 +670,70 @@ func (devices *DeviceSet) loadMetadata(hash string) *DevInfo {
 }
 
+func getDeviceUUID(device string) (string, error) {
+	out, err := exec.Command("blkid", "-s", "UUID", "-o", "value", device).Output()
```
is there simple way to do this without shelling out?

rhvgoyal replied:
I am not aware of any other easy way to do it. blkid comes from util-linux package and mkfs comes from util-linux too. As we are already using mkfs, I think it is safe to assume that blkid is available too on machines running docker.

vbatts replied:
there is a header and way to link to libbklid, but that seems too much. Please would require additional build time packages, where `blkid` is provided by `util-linux` which is on every major distro (even debian minbase).

vbatts commented:
LGTM

vbatts commented:
@tiborvass care to review?

calavera commented on daemon/graphdriver/devmapper/deviceset.go:
```diff
@@ -721,6 +783,7 @@ func (devices *DeviceSet) setupBaseImage() error {
 		return err
 	}
 
+	devices.saveBaseDeviceUUID(info)
```
is the error returned by `saveBaseDeviceUUID` ignored here on purpose? should we return it?

vbatts replied:
Not critical, but wouldn't hurt
On Jun 16, 2015 17:40, "David Calavera" <notifications@github.com> wrote:

> In daemon/graphdriver/devmapper/deviceset.go
> <https://github.com/docker/docker/pull/13896#discussion_r32574282>:
>
> > @@ -721,6 +783,7 @@ func (devices *DeviceSet) setupBaseImage() error {
> >  		return err
> >  	}
> >
> > +	devices.saveBaseDeviceUUID(info)
>
> is the error returned by saveBaseDeviceUUID ignored here on purpose?
> should we return it?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/docker/docker/pull/13896/files#r32574282>.
>

rhvgoyal replied:
Ok I added the error check and pushed the patch again. PTAL.

rhvgoyal committed:
devicemapper: Compare uuid of base device on startup

It is easy for one to use docker for a while, shut it down and restart
docker with different set of storage options for device mapper driver
which will effectively change the thin pool. That means any of the
metadata stored in /var/lib/docker/devicemapper/metadata/ is not valid
for the new pool and user will run into various kind of issues like
container not found in the pool etc.

Users think that their images or containers are lost but it might just
be the case of configuration issue. People might use wrong metadata
with wrong pool.

To detect such situations, save UUID of base image and once docker
starts later, query and compare the UUID of base image with the
stored one. If they don't match, fail the initialization with the
error that UUID failed to match.

That way user will be forced to cleanup /var/lib/docker/ directory
and start docker again.

Signed-off-by: Vivek Goyal <vgoyal@redhat.com>

calavera commented:
LGTM. Merging :confetti_ball:
